src/Common/src/System/Security/Cryptography/Asn1Reader/AsnWriter.BitString.cs

```diff
@@ -120,20 +117,190 @@ private void WriteBitStringCore(Asn1Tag tag, ReadOnlySpan<byte> bitString, int u
+            }
+
+            // T-REC-X.690-201508 sec 8.6.2.2
+            if (unusedBitCount < 0 || unusedBitCount > 7)
+            {
+                throw new ArgumentOutOfRangeException(
+                    nameof(unusedBitCount),
+                    unusedBitCount,
+                    SR.Cryptography_Asn_UnusedBitCountRange);
+            }
+
+            CheckDisposed();
+
+            int savedOffset = _offset;
+            Span<byte> scratchSpace;
+            byte[] ensureNoExtraCopy = null;
+            int expectedSize = 0;
+
+            // T-REC-X.690-201508 sec 9.2
+            //
+            // If it's not within a primitive segment, use the constructed encoding.
+            // (>= instead of > because of the unused bit count byte)
+            bool segmentedWrite =
+                RuleSet == AsnEncodingRules.CER && byteLength >= AsnReader.MaxCERSegmentSize;
+
+            if (segmentedWrite)
+            {
+                // Rather than call the callback multiple times, grow the buffer to allow
+                // for enough space for the final output, then return a buffer where the last segment
+                // is in the correct place. (Data will shift backwards to the right spot while writing
+                // other segments).
+                expectedSize = DetermineCerBitStringTotalLength(tag, byteLength);
+                EnsureWriteCapacity(expectedSize);
+                int overhead = expectedSize - byteLength;
+
+                // Start writing where the last content byte is in the correct place, which is
+                // after all of the overhead, but ending before the two byte end-of-contents marker.
+                int scratchStart = overhead - 2;
+                ensureNoExtraCopy = _buffer;
+                scratchSpace = _buffer.AsSpan(scratchStart, byteLength);
+
+                // Don't let gapped-writes be unpredictable.
+                scratchSpace.Clear();
+            }
+            else
+            {
+                WriteTag(tag.AsPrimitive());
+                // The unused bits byte requires +1.
+                WriteLength(byteLength + 1);
+
+                _buffer[_offset] = (byte)unusedBitCount;
+                _offset++;
+
+                scratchSpace = _buffer.AsSpan(_offset, byteLength);
+            }
+
+            action(scratchSpace, state);
+
+            // T-REC-X.690-201508 sec 11.2
+            //
+            // This could be ignored for BER, but since DER is more common and
+            // it likely suggests a program error on the caller, leave it enabled for
+            // BER for now.
+            if (!CheckValidLastByte(scratchSpace[byteLength - 1], unusedBitCount))
+            {
+                // Since we are restoring _offset we won't clear this on a grow or Dispose,
+                // so clear it now.
+                _offset = savedOffset;
+                scratchSpace.Clear();
+
+                // TODO: Probably warrants a distinct message.
```

Comment:
Are these TODOs tracked somewhere?

Reply:
Just with their existence :smile:.  It's mainly notes for when the types become public, and should get a bit friendlier.